pkg/apis/networking/validation/validation.go

```diff
@@ -597,3 +598,78 @@ func allowInvalidWildcardHostRule(oldIngress *networking.Ingress) bool {
 	}
 	return false
 }
+
+// ValidateClusterCIDRConfigName validates that the given name can be used as an
+// ClusterCIDRConfig name.
+var ValidateClusterCIDRConfigName = apimachineryvalidation.NameIsDNSLabel
+
+// ValidateClusterCIDRConfig validates a clusterCIDRConfig.
+func ValidateClusterCIDRConfig(ccc *networking.ClusterCIDRConfig) field.ErrorList {
+	allErrs := apivalidation.ValidateObjectMeta(&ccc.ObjectMeta, false, ValidateClusterCIDRConfigName, field.NewPath("metadata"))
+	allErrs = append(allErrs, ValidateClusterCIDRConfigSpec(&ccc.Spec, field.NewPath("spec"))...)
+	return allErrs
+}
+
+// ValidateClusterCIDRConfigSpec validates clusterCIDRConfig Spec.
+func ValidateClusterCIDRConfigSpec(spec *networking.ClusterCIDRConfigSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, unversionedvalidation.ValidateLabelSelector(spec.NodeSelector, fldPath.Child("nodeSelector"))...)
+
+	maxIPv4PerNodeMaskSize := int32(32)
+	maxIPv6PerNodeMaskSize := int32(128)
+
+	hasIPv4 := spec.IPv4 != nil
+	hasIPv6 := spec.IPv6 != nil
+
+	// Validate if CIDR is configured for at least one IP Family(IPv4/IPv6).
+	if !hasIPv4 && !hasIPv6 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("ipv4", "ipv6"), *spec, "either IPv4 or IPv6 CIDR must be configured, both cannot be empty"))
+	}
+
+	// Validate configured IPv4 CIDR and PerNodeMaskSize.
+	if hasIPv4 {
+		if !netutils.IsIPv4CIDRString(spec.IPv4.CIDR) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("ipv4", "cidr"), *spec.IPv4, "provided CIDR is not a valid IPv4 CIDR"))
+		}
+		if spec.IPv4.PerNodeMaskSize < 0 || spec.IPv4.PerNodeMaskSize > maxIPv4PerNodeMaskSize {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("ipv4", "perNodeMaskSize"), *spec.IPv4, "provided IPv4 PerNodeMaskSize is invalid, Valid values: [0,32]"))
+		}
```

Comment:
the max perNodeMask size depends on the CIDR mask, doesn't it?

if spec.IPv4.CIDR 10.0.0.0/8  it has to be between 8 and 32

Reply:
That's correct, updated the validation check